### DISCUSSION TITLE
Inserting images

### DISCUSSION THREAD

**Discussion by CarlosR2** [vkbo/novelWriter#801] [Q&A] [answered]
I just downloaded and was toying with the app, btw, I really like it and i find it very useful.

But i was wondering if you could insert pictures, i haven't seen the option.

Thanks

**Comment by vkbo** [comment, marked as answer]
I'm glad you like it!

novelWriter is a plain text editor intended for writing pure text. It has no support for images, links, coloured text, tables or any other such structures. Just plain headers and paragraphs. The colouring that is added is purely for visual aid, and isn't saved to the text files.

**Reply by CarlosR2** [reply]
Thanks for your answer Veronica.

I just thought would be nice to have the ability to add images since many books include figures. I wish i had enough knowledge in python to contribute.

And congrats again!

**Reply by vkbo** [reply]
Oh, it's not a matter of being able to do it. Omitting general word processor features is intentional. It just isn't what the editor is for. It's for writing fiction by assisting in the process of writing – and only writing. Its end product is a manuscript document. I'm still planning to add features, but only features that fit within that scope.

Novels do not generally have figures. Some have illustrations, however they are usually not a part of the text but rather fluff added by the publisher for special editions and such. In any case, they're not a part of the writing process. Even when self-publishing, you'd want to manually post-process the final manuscript into something more presentable and reader-ready.

There are plenty of writing tools out there with a broader range of use. I have used them myself, and I made novelWriter exactly because I wanted something more tailored to writing fiction and not have all that extra stuff I don't need, and which just gets in the way.

So yeah, the current range of features is very much deliberate :)

**Reply by CarlosR2** [reply]
good thinking! 
clarity is key

**Reply by longqzh** [reply]
@CarlosR2 maybe you should try `VNote`, it supports image, math formulas, table, UML and so on :)

**Comment by snden** [comment]
I would like to once again open a certain possibility regarding images. I know, I've read it here several times, that images are not and will not be part of Novelwriter. But I offer another point of view and ask for your consideration. When I wrote a book years ago, pictures were an integral part of the text. The publisher wanted clean, unformatted text (that's what Novelwriter is for, and thanks for keeping it clean, it's its unique feature, please keep it in the future), but he wanted that I insert to the text, where the image would be, specifically marked text. For example:
*** Figure 1 - Image Caption ***
I think that just such a simple text mark would help to have an overview of where the image will be in the final layout and would not disturb the author with graphics. I would suggest, for example, the tag \@ picture: , and then a special folder called pictures in the tree of the book structure (similarly as for Characters, Objects ...) where could be written detailed notes about the image in its corresponding file, a link to it (path where it is stored), its detailed description ... . During the final creation and export of the book, if it were possible to renumber this 'links' to the pictures chronologically, it would save would be a lot of work for the author, because he often moves and deletes chapters. I had dozens of pictures in the book, and without marking them directly in the text right when I was writing, I would get lost overview in where each one should be placed. Sorry for re-opening the image topic. Perhaps this will be a more acceptable way to incorporate images into the text, as it will only be a short textual representation of it.

Just to explain what i mean exactly - in the resulting text generated by the Build Novel Project, there would not be an image in the text, but only its textual representation. Just *** Figure 1 - Image Caption *** (Or any other wildcard format that the publisher would require.), the typesetter would then deal with that when typesetting of the book. The result of compiling the entire text of the book would be, for example:

  Lorem ipsum dolor sit amet, consectetuer adipiscing elit.
*** Figure 1 - Image1 Caption ***
Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo.
*** Figure 2 - Image2 Caption ***
Sed convallis magna eu here.

**Reply by vkbo** [reply]
Ok, let me be a bit wordy, since this is a discussion topic, and I would also like to sum up my own thinking so far.

I have been considering something like this. It ties into a few other requested features that may be combined into a set of similar extensions.

Firstly, some concerns: novelWriter is based on Markdown in principle, but only very loosely. By design the markup doesn't allow for advanced structures, which is something the reStructureText markup system solves. Extending the novelWriter syntax to allow more complex structures will require very careful considerations as it is difficult to alter it later on. The syntax also needs to be consistent and logical.

So, at present, we have the following styles:

* Headers are denoted by `#` symbols, sometimes modified by a `!`. That's a minor extension to Markdown, and the `!` only matters in terms of formatting, not structure.
* We have Markdown style text emphasis.
* The `@` symbol is used for meta data about text sections, which is unique. The `@` syntax should not be used for other things, as it would be confusing.
* Lines starting with `%` are considered to be not part of the text itself. There is already an extension to comments for synopsis.
* The `>` and `<` can be used to align or block indent text. I took the idea from the plain text quote format from emails, and it is also used by Markdown, although novelWriter treats it slightly differently.
* Some special formatting commands exist using a `[COMMAND:VALUE]` notation. These are used to insert special formatting for page layout and are generally not needed in the text.

These are the base formatting principles. It is already quite a bit, and some of it is unique to novelWriter, so it adds an additional burden on the user to learn new syntax.

In addition, we have the following requests that would need a plain text syntax:
* #342
* #399
* #705 
* #1050
* #1060 
* #1075 
* There's also been discussion about syntax to raise and lower text.

Now, some of those are not really achievable without adding very clunky notation. It also poses challenges to the syntax highlighting feature which may need a complete rewrite for the more complex requests.

In order to add new features, I would very much like to settle on an extendable syntax, ideally based on one of the existing code sets. I need to settle on a method for inserting stuff into the text, like for footnotes, and raised and lowered text. I would also like to extend the syntax of comments to allow for differentiating between types of the writer's notes that isn't part of the novel text. On that note, I propose:

* Any text line starting with `%` is always ignored by the text scanning tools and excluded from the body text of the novel.
* Any modifiers to the plain comment should follow a similar syntax as the `% synopsis:` one.
* I want to add a feature for outtakes that isn't just plain comments. I've considered making it `%~`, which is consistent with the in-line syntax for deleted text, `~~`.

**So, finally, to your request**

I don't want to add a `@picture` or `@figure` keyword, as they would not work like the other `@` keywords. However, if the feature you're requesting could be generalised to work for any editorial note to be exported into the manuscript, maybe we can find a sensible solution. 

As it is now, you can already leave comments in the manuscript. However, a lot of people use comments for text taken out of the story, or to add information only relevant to the writer – a note to self sort of thing – which I do all the time by the way.

It could be a solution to allow the user to flag certain comments as a manuscript note, and let it be exported separately from other comments, just like the synopsis is. In order to use it for figures, you could add your own searchable keyword to the comments as comments are not parsed by the build engine of novelWriter. They are just converted to plain text paragraphs.

Proposed syntax:
```
% note: Some text to appear as a note in the manuscript.
```

It is consistent with existing notation, can be formatted specially in exports, and filtered in or out as needed, and highlighted in the editor in a similar way as synopsis comments are. They can also be separately parsed by the indexer using the existing logic that exists for synopsis comments with very little processing overhead.

Since they will be indexed, it could also be possible to show the location of these notes in the Outline view, and make then readable there. It wouldn't be exactly like the synopsis comments, but it wouldn't be difficult to make them easy to locate.

Now, in order to use them for figures, you could just format them accordingly, e.g.:

```
% note: Figure: Some text describing the figure and a reference to the name of the relevant file.
```
If you really need to tag them in documents as well, the `@custom` tag can achieve this easily. Just name the tags with a `Figure` prefix to separate them from other custom elements, if you have them. You can make a custom root folder named "Figures" as well, and keep the project notes separate.

This way, there is no new narrow functionality that needs to be added, and the note feature could be generally useful as a custom feature that the writer can use for multiple purposes.

What do you think?

**Reply by snden** [reply]
Thank you for the detailed description. According to this, I realize that keywords starting with @ have a really specific function regarding the text. I've noticed that they also behave in a specific way, such as when written directly in text, they being placed always at the beginning. This would lose the ability to show where to place the image. It's a good idea to create a general manuscript note option. This could be generated or not when the resulting file is rendered, independent of the regular comment.
Just my idea is to separate the simple note from the note referring to the image, so that when you disable the notes when generating the output, the image references don't disappear and remain in the text. That is, the format
% note: Figure: Some text describing the figure and a reference to the name of the relevant file.
is a solution, moreover, it will be possible to use it for a certain form of communication with the typewriter. But it has the only drawback that I can think of, that it will not be possible to create a list of images used in the document. That is, to have a overview of which image was included in which chapters, as is common for Objects, Characters ...
Using such an unusual thing for markdown as an html tag in the style of <img>Description of image</img>, or without </img> if it was on a single line, is probably a bad idea of me, because it would no longer be markdown syntax. My thoughts just flow like that. I like Markdown for its simplicity. This should not be lost on the Novelwriter.
I can think of some alternative of the note or comment, for example using a backslash
\ note:
\ picture: Description of picture
or as comments in some programming languages
/* picture: Description of picture
(no */ trailing character)
or
// picture: Description of picture

This would distinguish, that % the note would be unprintable
and /* or // the note would be printable and, in the case of an image, perhaps even numberable, because it would be not just a simple comment.

I'm glad we can discuss this because images are certainly a frequent request and I think this option would bring Novelwriter closer to many more users who are still hesitant to fully commit to it for serious work.

I'll still thinking about.

**Reply by vkbo** [reply]
You don't have to add the `@` keywords at the beginning. As long as they're on their own line, they can be added anywhere.

In any case, my suggestion on using the `@custom` tag was precisely to indicate which chapter/scene has which figure. It would then show up in the Outline. The placement of the image would be indicated by the note, which would also provide the description.

As far as I can see, these two points would give you what you need, and when a way to list references, as discussed in #1127, is added, you'll have that too. I'll look into adding a way to generate a complete ToC for the manuscript in the new build tool for release 1.8. That would make it easier to extract this info.

**Reply by vkbo** [reply]
Here's an example using the sample project, adding two figures:

1. Create a custom root folder, and add a note for each figure
![image](https://user-images.githubusercontent.com/1619840/193461092-1ca03e39-7498-4b01-be6d-4ad09ad6c05a.png)

2. Tag the figures where you want them to show up, using the reference keyword `@custom:`
![image](https://user-images.githubusercontent.com/1619840/193461127-07ad940e-af08-49d4-a0a4-a25b9c3b8c41.png)

3. The location of the figures will show up in the Custom column in the Outline View.
![image](https://user-images.githubusercontent.com/1619840/193461147-64b486d7-ab21-43c9-bfee-5d94c0a49e89.png)

**Reply by snden** [reply]
If it was possible to separate 'Include keywords' and 'include custom keyword' when generating the resulting file, that would be a solution.

**Reply by vkbo** [reply]
Hmm, why don't I add the option to select which keywords to export during the build? I'm rewriting the entire tool anyway, so may as well add that in the process.

**Reply by vkbo** [reply]
I've added feature tickets #1132 and #1133 for the 1.8 release cycle.

**Reply by snden** [reply]
Excellent! I wanted to suggest one more improvement, but I didn't want to bother you with it anymore. It has to do with metadata when exporting. Would it be possible to distinguish the metadata from each other so that it has a different style in the exported file? In . odt - style, in .html - .class? Then it would be possible to work with the metadata in another editor (change font size, color, hide...)

for example in html: 
\<span class="Locations"\>
\<span class="Characters"\>
\<span class="POV"\>
\<span class="Timeline"\>
\<span class="Objects"\>
\<span class="Focus"\>
\<span class="Synopsis"\>
\<span class="Comment"\>
\<span class="Tag"\>

**Reply by vkbo** [reply]
Sure, that's fairly easy. Just make a feature request and I'll add it to the 1.8 backlog.

Edit: It's easy for HTML anyway. I have to look into it for ODT.
